src/frequenz/sdk/actor/power_distributing/result.py

```diff
@@ -3,122 +3,92 @@
-                succeed.
-            failed_power: Part of the requested power that failed.
-            failed_batteries: Subset of the requested batteries for which the request
-                failed.
-            excess_power: Part of the requested power that could not be fulfilled,
-                because it was outside available power bounds.
-        """
-        super().__init__(request)
-        self.succeed_power: int = succeeded_power
-        self.succeed_batteries: Set[int] = succeeded_batteries
-        self.failed_power: int = failed_power
-        self.failed_batteries: Set[int] = failed_batteries
-        self.excess_power: int = excess_power
-
-
-class Error(Result):
-    """Error occurred and power was not set."""
-
-    def __init__(self, request: Request, msg: str) -> None:
-        """Create class instance.
-
-        Args:
-            request: The user's request to which this message responds.
-            msg: Error message explaining why error happened.
-        """
-        super().__init__(request)
-        self.msg: str = msg
-
-
-class OutOfBound(Result):
-    """Send if power was not set because requested power was not within bounds.
-
-    This message is send only if Request.adjust_power = False.
+@dataclasses.dataclass
+class _BaseResultMixin:
+    """Base mixin class for reporting power distribution results."""
+
+    request: Request
+    """The user's request to which this message responds."""
+
+
+# When moving to Python 3.10+ we should replace this with an union type:
+# Result = Success | PartialFailure | Error | OutOfBound | Ignored
+# For now it can't be done because before 3.10 isinstance(result, Success)
+# doesn't work, so it is hard to figure out what type of result you got in
+# a forward compatible way.
+@dataclasses.dataclass
+class Result(_BaseResultMixin):
+    """Power distribution result."""
+
+
+@dataclasses.dataclass
+class _BaseSuccessMixin:
+    """Result returned when setting the power succeed for all batteries."""
+
+    succeeded_power: int
+    """The part of the requested power that was successfully set."""
+
+    succeeded_batteries: set[int]
+    """The subset of batteries for which power was set successfully."""
+
+    excess_power: int
+    """The part of the requested power that could not be fulfilled.
+
+    This happens when the requested power is outside the available power bounds.
+    """
+
+
+@dataclasses.dataclass
+class Success(Result, _BaseSuccessMixin):
+    """Result returned when setting the power succeed for all batteries."""
+
+
+@dataclasses.dataclass
+class PartialFailure(Result, _BaseSuccessMixin):
+    """Result returned when any battery failed to perform the request."""
+
+    failed_power: int
+    """The part of the requested power that failed to be set."""
+
+    failed_batteries: set[int]
+    """The subset of batteries for which the request failed."""
+
+
+@dataclasses.dataclass
+class Error(Result, _BaseResultMixin):
```

Comment:
`Result` already derives from `_BaseResultMixin`,  do we need to derive from it separately also?

Reply:
I guess not, it was probably an artifact from my attempt to use `Result = Success | ...`. I can remove it for now.

Reply:
Done and updated the PR.